### PR TITLE
Properly handle multiple metadata entries that have the same key

### PR DESCRIPTION
**Pull request by NebelNidas**
The previous behavior was inconsistent, now it properly supports multiple entries with equal keys. If only the latest entry per key is desired, the `NEEDS_METADATA_UNIQUENESS` flag has to be passed.

### CODE REVIEW

**Review by modmuss50 — APPROVED** [review]

**Comment by liach** [comment]
I think the only problem with the previous version is that it cannot remove all entries with the same key. You can use `Collection.removeIf` and return the first value removed. Alternatively we can just return a modifiable collection for `getMetadata` and deprecate `String remove(String)`.

**Comment by NebelNidas** [comment]
I can't think of a case where you want to allow multiple metadata entries with the same key. @sfPlayer1 What's your opinion on this?

**Review by modmuss50 — APPROVED** [review]

**Review by modmuss50 — APPROVED** [review]

File: src/main/java/net/fabricmc/mappingio/format/tiny/Tiny1FileWriter.java [hunk 1]
```diff
@@ -190,7 +190,7 @@ private void writeTab() throws IOException {
 		writer.write('\t');
 	}
 
-	private static final Set<MappingFlag> flags = EnumSet.of(MappingFlag.NEEDS_SRC_FIELD_DESC, MappingFlag.NEEDS_SRC_METHOD_DESC);
+	private static final Set<MappingFlag> flags = EnumSet.of(MappingFlag.NEEDS_METADATA_UNIQUENESS, MappingFlag.NEEDS_SRC_FIELD_DESC, MappingFlag.NEEDS_SRC_METHOD_DESC);
```
[the remark this answers is not in the document]
Comment (NebelNidas) [reply src/main/java/net/fabricmc/mappingio/format/tiny/Tiny1FileWriter.java:193]: Okay, going to remove it then 👍

File: src/main/java/net/fabricmc/mappingio/tree/MemoryMappingTree.java [hunk 2]
```diff
@@ -303,8 +292,24 @@ public void accept(MappingVisitor visitor, VisitOrder order) throws IOException
 		do {
 			if (visitor.visitHeader()) {
 				visitor.visitNamespaces(srcNamespace, dstNamespaces);
+				List<MetadataEntry> metadataToVisit = metadata;
+
+				if (visitor.getFlags().contains(MappingFlag.NEEDS_METADATA_UNIQUENESS)) {
+					metadataToVisit = new LinkedList<>();
```
[the remark this answers is not in the document]
Comment (NebelNidas) [reply src/main/java/net/fabricmc/mappingio/tree/MemoryMappingTree.java:298]: I'm doing lots of front-inserts, isn't a linked list better for that?

File: src/main/java/net/fabricmc/mappingio/tree/MemoryMappingTree.java [hunk 3]
```diff
@@ -1699,6 +1705,44 @@ public String toString() {
...
+
+			MetadataEntryImpl entry = (MetadataEntryImpl) other;
+
+			return this.key.equals(entry.key) && this.value.equals(entry.value);
+		}
+
+		@Override
+		public int hashCode() {
+			return key.hashCode() | value.hashCode();
+		}
```
[the remark this answers is not in the document]
Comment (NebelNidas) [reply src/main/java/net/fabricmc/mappingio/tree/MemoryMappingTree.java:1740]: Would `return key + ":" + value;` work?

File: src/main/java/net/fabricmc/mappingio/tree/MemoryMappingTree.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (liach) [reply src/main/java/net/fabricmc/mappingio/tree/MemoryMappingTree.java:298]: You can declare `metadataToVisit` as a `Collection` and use `ArrayDeque` to add to head of a collection quickly.

**Pull request merged**